### PR TITLE
Dockerfile for Shotover

### PR DESCRIPTION
**Pull request by conorbros**
Example of how to use, assuming your config and topology file are in the home dir:
`docker run -t -v /home/conor:/config shotover-proxy --help`
`docker run -t -v /home/conor:/config shotover-proxy -t config/top1.yaml -c config/config.yaml`

I've registered a "shotover" user on Dockerhub and will also integrate publishing our releases in this PR.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM rust:latest as builder
+
+WORKDIR ./shotover-proxy
+
+COPY ./ ./
+
+# Clean out anything already built in the target dir
+RUN rm -rf target
```
Comment (XA21X) [inline Dockerfile:8]: You can add a `.dockerignore` file to exclude paths from the build context (just like `.gitignore`).

**Comment by conorbros** [comment]
I've tested the workflow on this repo: https://github.com/conorbros/hot-restart-proxy
Published results can be seen here: https://hub.docker.com/repository/docker/conorbros/hot-restart-proxy

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
+FROM rust:latest as builder
+
+WORKDIR ./shotover-proxy
+
+COPY ./ ./
+
+RUN cargo build
```
Comment (rukai) [inline Dockerfile:7]: I was wondering why the binary was 239MB,
you forgot to build in `--release`!

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+FROM debian:bullseye-slim
+
+RUN apt-get update -y \
+    && apt-get install glibc-doc -y \
```
Comment (rukai) [inline Dockerfile:12]: why do we install glibc-doc?

**Review by rukai — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 3 above)
Comment (conorbros) [reply Dockerfile:12]: Not on `debian:bulleye-slim` by default.

File: Dockerfile (hunk 2 above)
Comment (conorbros) [reply Dockerfile:7]: Oops, yep I didn't want to wait as long in between tests.

File: Dockerfile (hunk 3 above)
Comment (rukai) [reply Dockerfile:12]: what is it though?

Comment (conorbros) [reply Dockerfile:12]: Was a mistake leaving it in. Removed it as `debian:bullseye` includes by default.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+WORKDIR /shotover-proxy
+
+COPY ./ ./
+
+RUN cargo build --release
+
+FROM debian:bullseye-slim
+
+RUN apt-get update -y \
```
Comment (rukai) [inline Dockerfile:11]: isnt this basically a no-op?
updating the lists and then deleting them?
Or is there a reason for it?

Comment (conorbros) [reply Dockerfile:11]: Oops, there should be an upgrade in there.

Comment (rukai) [reply Dockerfile:11]: Is that a normal thing to do?
Wouldnt the debian image already have up to date packages?
Looking at redis dockerfile I can see that it doesnt run upgrade https://github.com/docker-library/redis/blob/master/6.2/Dockerfile

Comment (rukai) [reply Dockerfile:11]: https://github.com/docker-library/cassandra/blob/master/4.0/Dockerfile cassandra too

Comment (conorbros) [reply Dockerfile:11]: Ok sure I'll leave out the upgrading.

**Review by rukai — APPROVED** [review]
I dont have much experience with docker but this all seems reasonable to me:
* `docker image save shotover -o foo.tar.gz` shows nothing strange is being accidentally included.
* `docker run --network host -p 6379:6379 -v $PWD/config:/config shotover` runs fine off the default config files in the shotover repo and I can succesfully send a redis message.

**Review by XA21X — APPROVED** [review]

**Pull request merged**